File: frontend/src/scene/vr_freeform_placement.js
/** Source-space rigid placement, never tracking metres or a display scale. */
export function normalizeFreeformPlacement(value) {
  if (value === undefined) return undefined
  if (!value || Array.isArray(value) || typeof value !== 'object' ||
      Object.keys(value).sort().join(',') !== 'rotation_xyzw,translation_nm') return null
  const { translation_nm:t,rotation_xyzw:q } = value
  if (!Array.isArray(t) || t.length!==3 || !t.every(v=>Number.isFinite(v) && Math.abs(v)<=1e6) ||
      !Array.isArray(q) || q.length!==4 || !q.every(Number.isFinite)) return null
  const norm = Math.hypot(...q)
  if (Math.abs(norm-1)>1e-3) return null
  return { translation_nm:[...t],rotation_xyzw:q.map(v=>v/norm) }
}

File: backend/core/vr_freeform_placement.py
"""Bounded rigid source-coordinate placement for native extrusion drafts."""
import math


def validate_freeform_placement(value):
    if not isinstance(value, dict) or set(value) != {'translation_nm', 'rotation_xyzw'}:
        raise ValueError('invalid freeform placement')
    t, q = value['translation_nm'], value['rotation_xyzw']
    def vector(v, size):
        return (isinstance(v, list) and len(v) == size and
                all(type(n) in (int, float) and math.isfinite(n) for n in v))
    if not vector(t, 3) or any(abs(n) > 1e6 for n in t) or not vector(q, 4):
        raise ValueError('invalid freeform placement vector')
    norm = math.hypot(*q)
    if abs(norm-1) > 1e-3:
        raise ValueError('freeform placement requires a unit quaternion')
    return {'translation_nm':list(t), 'rotation_xyzw':[n/norm for n in q]}
